swarm_test.go

```diff
@@ -280,60 +282,102 @@ func TestConnHandler(t *testing.T) {
+
+		p1ConnectednessToP2 network.Connectedness
+		p2ConnectednessToP1 network.Connectedness
+		isP1OutboundErr     bool
+	}{
+		"no gating": {
+			p1ConnectednessToP2: network.Connected,
+			p2ConnectednessToP1: network.Connected,
+			isP1OutboundErr:     false,
+		},
+		"p1 gates outbound peer dial": {
+			p1Gater: func(c *MockConnectionGater) *MockConnectionGater {
+				c.PeerDial = func(p peer.ID) bool { return false }
+				return c
+			},
+			p1ConnectednessToP2: network.NotConnected,
+			p2ConnectednessToP1: network.NotConnected,
+			isP1OutboundErr:     true,
+		},
+		"p1 gates outbound addr dialing": {
+			p1Gater: func(c *MockConnectionGater) *MockConnectionGater {
+				c.Dial = func(p peer.ID, addr ma.Multiaddr) bool { return false }
+				return c
+			},
+			p1ConnectednessToP2: network.NotConnected,
+			p2ConnectednessToP1: network.NotConnected,
+			isP1OutboundErr:     true,
+		},
+		"p2 gates inbound peer dial before securing": {
+			p2Gater: func(c *MockConnectionGater) *MockConnectionGater {
+				c.Accept = func(c network.ConnMultiaddrs) bool { return false }
+				return c
+			},
+			p1ConnectednessToP2: network.NotConnected,
+			p2ConnectednessToP1: network.NotConnected,
+			isP1OutboundErr:     true,
+		},
+		"p2 gates inbound peer dial before multiplexing": {
+			p1Gater: func(c *MockConnectionGater) *MockConnectionGater {
+				c.Secured = func(network.Direction, peer.ID, network.ConnMultiaddrs) bool { return false }
+				return c
+			},
+			p1ConnectednessToP2: network.NotConnected,
+			p2ConnectednessToP1: network.NotConnected,
+			isP1OutboundErr:     true,
+		},
+		"p2 gates inbound peer dial after upgrading": {
+			p1Gater: func(c *MockConnectionGater) *MockConnectionGater {
+				c.Upgraded = func(p transport.CapableConn) (bool, control.DisconnectReason) { return false, 0 }
+				return c
+			},
+			p1ConnectednessToP2: network.NotConnected,
+			p2ConnectednessToP1: network.NotConnected,
+			isP1OutboundErr:     true,
+		},
+		"p2 gates outbound dials": {
+			p2Gater: func(c *MockConnectionGater) *MockConnectionGater {
+				c.PeerDial = func(p peer.ID) bool { return false }
+				return c
+			},
+			p1ConnectednessToP2: network.Connected,
+			p2ConnectednessToP1: network.Connected,
+			isP1OutboundErr:     false,
+		},
 	}
 
-	swarms[0].Peerstore().AddAddr(swarms[1].LocalPeer(), swarms[1].ListenAddresses()[0], peerstore.PermanentAddrTTL)
-	_, err = swarms[0].DialPeer(ctx, swarms[1].LocalPeer())
-	if err == nil {
-		t.Fatal("dial should have failed")
-	}
-}
-
-func TestFilterBounds(t *testing.T) {
-	ctx := context.Background()
-	swarms := makeSwarms(ctx, t, 2)
-
-	conns := make(chan struct{}, 8)
-	swarms[0].SetConnHandler(func(conn network.Conn) {
-		conns <- struct{}{}
-	})
+	for n, tc := range tcs {
+		p1Gater := DefaultMockConnectionGater()
+		p2Gater := DefaultMockConnectionGater()
+		if tc.p1Gater != nil {
+			p1Gater = tc.p1Gater(p1Gater)
+		}
+		if tc.p2Gater != nil {
+			p2Gater = tc.p2Gater(p2Gater)
+		}
```

Comment:
Can we use `t.Run()` here to make these actual subtests instead of one huge test with many cases?

Reply:
Good idea. Done.